haystack/nodes/reader/table.py

```diff
@@ -275,16 +389,187 @@ def _predict_tapas_for_qa(self, inputs: BatchEncoding, document: Document) -> An
+        all_cell_probabilities = self.tokenizer._get_mean_cell_probs(
+            token_probabilities[0].tolist(), segment_ids, row_ids, column_ids
+        )
+        # _get_mean_cell_probs seems to index cells by (col, row). DataFrames are, however, indexed by (row, col).
+        all_cell_probabilities = {(row, col): prob for (col, row), prob in all_cell_probabilities.items()}
+        answer_cell_probabilities = [all_cell_probabilities[coord] for coord in answer_coordinates]
+
+        return np.mean(answer_cell_probabilities)
+
+    @staticmethod
+    def _aggregate_answers(agg_operator: Literal["COUNT", "SUM", "AVERAGE"], answer_cells: List[str]) -> str:
+        if agg_operator == "COUNT":
+            return str(len(answer_cells))
+
+        # No aggregation needed as only one cell selected as answer_cells
+        if len(answer_cells) == 1:
+            return answer_cells[0]
+        # Return empty string if model did not select any cell as answer
+        if len(answer_cells) == 0:
+            return ""
+
+        # Parse answer cells in order to aggregate numerical values
+        parsed_answer_cells = [parser.parse(cell) for cell in answer_cells]
+        # Check if all cells contain at least one numerical value and that all values share the same unit
+        try:
+            if all(parsed_answer_cells) and all(
+                cell[0].unit.name == parsed_answer_cells[0][0].unit.name for cell in parsed_answer_cells
+            ):
+                numerical_values = [cell[0].value for cell in parsed_answer_cells]
+                unit = parsed_answer_cells[0][0].unit.symbols[0] if parsed_answer_cells[0][0].unit.symbols else ""
+
+                if agg_operator == "SUM":
+                    answer_value = sum(numerical_values)
+                elif agg_operator == "AVERAGE":
+                    answer_value = mean(numerical_values)
+                else:
+                    raise ValueError("unknown aggregator")
+
+                return f"{answer_value}{' ' + unit if unit else ''}"
+
+        except ValueError as e:
+            if "unknown aggregator" in str(e):
+                pass
+
+        # Not all selected answer cells contain a numerical value or answer cells don't share the same unit
+        return f"{agg_operator} > {', '.join(answer_cells)}"
+
+    def predict(self, query: str, documents: List[Document], top_k: int) -> Dict:
+        answers = []
+        table_documents = self._check_documents(documents)
+        for document in table_documents:
+            table: pd.DataFrame = document.content
+            model_inputs = self._preprocess(query, table)
+            model_inputs.to(self.device)
+
+            current_answer = self._predict_tapas(model_inputs, document)
+            answers.append(current_answer)
+
+        answers = self._postprocess(answers, top_k)
+        results = {"query": query, "answers": answers}
+        return results
+
+
+class _TapasScoredEncoder:
```

Comment:
Can we let `_TapasScoredEncoder` inherit from `_TapasEncoder`? There are some duplicate methods: 
- `_calculate_answer_offsets`
- `_check_documents`
- `_preprocess`

Reply:
Yes, I did notice that. I was worried that an inheritance would make the code a little hard to understand and that some of the functions in `_TapasEncoder` would not be relevant for `_TapasScoredEncoder` such as `_calculate_answer_score` and `_aggregate_answers` which I believe could cause confusion since those functions are not overwritten by `_TapasScoredEncoder`.

Two other considerations I can see:
1. make a `BaseTapasEncoder` that groups all common functionality together such as the ones you mentioned. I would also probably add `@abstractmethod` for `predict` to make it clear that whatever inhertis `BaseTapasEncoder` needs to implement a `predict` method.
2. Implement the functions you mentioned as free functions that are not in any class (e.g. util functions) such that we avoid duplicate code. 

What do you think?

Reply:
I see pros and cons for both options. I'd be fine with any of the two :)

Reply:
I went for option 1!